### ISSUE TITLE
Attempt to use 'Engines' section of package.json to set node version

### ISSUE THREAD

**Issue by fool** [netlify/build-image#53]
Not sure if this makes sense since the field can contain a range of versions rather than specifying an exact version, but maybe we could limit it to supporting only a definition for node that is a single version #?

This would be in addition to setting NODE_VERSION or .nvmrc since we obviously can't deprecate the existing methods without breaking a lot of builds.

what do you think @rybit ?

**Comment by rybit** [comment]
I am unfamiliar with that section of the package.json, but you're saying we could parse out a specific node version and try to set it via nvm? 

If this is the 'standard' (as much as anything in JS is standardized) then yes, but personally, I don't see this as being a great way. We'd have to use `jq` which isn't bad, but the range vs exact could be more confusing to end uses.

Not against it though.

**Comment by oskarrough** [comment]
Hi, I do believe `engines` in `package.json` is a kind of standard. In any case, it would be nice to avoid the extra nvm file if there's already a package.json.

https://yarnpkg.com/lang/en/docs/package-json/#toc-engines
https://docs.npmjs.com/files/package.json#engines
https://zeit.co/now#which-version-of-node.js-do-you-run

**Comment by RabidFire** [comment]
Hi! Just wanted to add that Heroku also uses the `engines` config in `package.json` to determine the node, npm & yarn versions. When ranges are used, I believe they simply use the best match. If nothing's provided, they use the latest LTS release:
https://devcenter.heroku.com/articles/nodejs-support

This tripped me up a bit when shifting my project from Heroku to Netlify. For now, I've created a `netlify.toml` file which duplicates my `engines` config, but it would be great to support this standard.

**Comment by bcomnes** [comment]
I would like to see this happen too.

**Comment by chris-hailstorm** [comment]
Having this in `package.json` would help keep multiple tools / services in sync.

**Comment by corysimmons** [comment]
Please have Netlify pay attention to package.json's `engines` object.

We should be able to:

```json
{
  "private": true,
  "engines": {
    "node": "14.4.0",
    "yarn": "1.22.4"
  }
}
```

and it just work.

Even after prefixing my `build` script with `NODE_VERSION=14.4.0 YARN_VERSION=1.22.4` I get version errors on Netlify. Do you guys _just_ support a few versions of Node/yarn???

Versions (specifically Node/npm/yarn) has been a goofy problem since Netlify was created and has probably cost you guys a ton of money in failed deploys. Please actually fix this once and for all with a simple solution (e.g. `NODE_VERSION=14.4.0 react-scripts build` actually works, and/or Netlify draws versions from the `engines` field (the more correct solution)).

---

Posterity: Netlify doesn't pull env vars out of commands. So `NODE_VERSION=14.4.0 react-scripts build` won't work. And they don't look at `engines` (they should). You need to add Node/yarn/npm versions here:

![image](https://user-images.githubusercontent.com/224636/84521782-8046ed00-aca3-11ea-92d6-5aa0c249648f.png)

**Comment by nolessafool** [comment]
As far as I know, our builds should support every released version of node  and yarn.  It is the specific inclusion of the `engines` line that may be problematic, or at best won't work.

If you'd like to link us to a failed deploy logs page once you've set `$NODE_VERSION` and `$YARN_VERSION` and removed the `engines` settings frompackage.json - probably best done by a [post in our community](https://community.netlify.com) rather than in this thread - our Support team will be happy to try to debug there.

This feature request is about supporting that line and is still good to follow for your use case, but I think we can come up with a workaround today - just not at the expense of everyone following this thread having to read our back and forth ;)

**Comment by corysimmons** [comment]
No worries. This is where I ended up from googling something like `netlify engines` so I figured I'd drop a screenshot for others.

**Comment by x5engine** [comment]
just set the NODE_VERSION: Sets the Node.js version.
on env variables
https://docs.netlify.com/configure-builds/environment-variables/#deploy-request-notifications

**Comment by yordis** [comment]
Any updates on respecting `engines.node`?

**Comment by micleners** [comment]
Pinging here again to see if this is moving to `package.json`.
I found this issue from this support forum: https://community.netlify.com/t/specifying-a-node-version/9701

**Comment by nolessafool** [comment]
thanks for the ping @micleners !  Any updates to this functionality will be posted in this thread :)  None are planned today as far as I know.  The workarounds listed here continue to be our best advice for today.

**Comment by ozum** [comment]
I had to spend a few hours understanding why my builds fail. Then later I saw, I had to set env variables for the node.js version. Although it is the only workaround currently available, it sure goes out of sync when we update "engines" in package.json and forget to update env in netlify.

EDIT: This is my first post, and I forgot to say thanks for providing such a good tool free for open source projects.

**Comment by quantuminformation** [comment]
Like this?

![image](https://user-images.githubusercontent.com/216566/114161825-c3f09e80-9928-11eb-818b-50ca042f10fa.png)

**Comment by quantuminformation** [comment]
I get this:
![image](https://user-images.githubusercontent.com/216566/114161993-f4383d00-9928-11eb-94ff-99cf573d5ddc.png)

**Comment by elwayman02** [comment]
Is there any movement on this feature request? `engines` is absolutely the standard way to define what node versions your project can be built with, so I'm not sure why we would want to manage it separately in Netlify. Not everyone uses NVM (some of us manage our node versions with `volta`, for example), so there won't always be a `.nvm` file to read the version from, but there will always be `package.json`, which already has the `engines` spec. It's really unexpected for Netlify to not know what version of Node to build with, when we've specified it in our project already.

**Comment by adamreisnz** [comment]
Yeah it would be really good if Netlify would just respect the semver Node version in the package.json engines definition instead of having to specify the version in two different places.

I don't see the fact that this version can be a range as a problem. You can simply apply the same version resolution algorithm as npm/yarn apply.

Meaning, if I specify Node version as "16.3.0" it will give me 16.3.0. If I specify it as "^16.3.0" it will give me whatever 16.x.x is available and the latest, but not 17, 18 etc.

**Comment by cesarvarela** [comment]
I'm surprised that some are asking if this is standard.

Anyway, yes it is.

edit. just realized that the `.node-version` file doesn't support semver, come on guys.

**Comment by ontoneio** [comment]
Can we please please please get this implemented?! It's so trivial and yet this issue has been open 4 years! 
This aspect of explicitly setting node versioning for Netlify still trips me up on new projects and it's becoming a nuisance at this point. Why is Netlify being so stubborn about this?

**Comment by tsongas** [comment]
So, somebody at Netlify didn't know about `engines` however many years ago so now all your customers are stuck with failing builds and workarounds? I don't get it, especially with the momentum toward TypeScript where your @types/node version is in `package.json`.

**Comment by MikeBeas** [comment]
Adding my support for this. I use this on Heroku and I don't use nvm at all because I don't need to manage multiple node versions on my laptop. All my projects are built with the latest versions. Netlify has been building my sites with the wrong npm and node versions (6.x and 12.x instead of 7.x and 16.x, respectively) and I only just now noticed.

This seems like a really obvious thing to support since it's part of the package.json standard. Seems wild it's been 5 years since it was requested without any movement.

**Comment by szamanr** [comment]
bumping this, it's pretty ridiculous that this has not been implemented yet.

**Comment by MadMaxMcKinney** [comment]
Once again sending this to the top. Got here after struggling to get the node version updated in an older project that got resurrected and spend a while figuring out why Netlify was failing to build. Is it still expected that we updated the node version via the environment var in the Netlify UI when needing to do so?

**Comment by braco** [comment]
Ah, the old spending 4 hours in Netlify build hell because of some weird unsupported thing again. Oof. Can anyone from Netlify take ownership of this problem? @erezrokah?

5 year old issue!

**Comment by MikeBeas** [comment]
I just ran into an issue where my builds were failing because I guess webpack requires something newer than Node 12, but that's what netlify was trying to build with despite my `engines` section specifying it should use Node ^17.0.0. This really needs to be addressed.

**Comment by MikeBeas** [comment]
Since there hasn't been any movement on this I went hunting to see how Heroku and others do it. Heroku is using a custom binary that checks their installable inventory on an S3 bucket and figures out what version matches the specified range, so that was kind of a non-starter, but there are other ways to get at this info.

`npm view node@^17.0.0 version | tail -n 1 | cut -d "'" -f2`

This will output the latest matching version of a package. In this case it's the latest version of node matching `^17.0.0`. It should work with any semver range you give it since npm handles that stuff. I tested with a bunch of them, including some unsafe ones like `<17.0.0` and `>17.0.0`. And, of course, it works with exact versions, too.

This solution isn't necessarily my favorite because it could require installing two versions of node/npm and only works with packages that are actually available from npm, but it _does_ work and removes any custom range parsing/logic concerns from your plate. Since most folks seem to be asking about node/npm/yarn, and those are all available on npm, it probably isn't the worst thing in the world to only support stuff available on npm.

In `install_dependencies()` after you've copied over the cached node version you could add a third `if` after the one that checks `.node-version` where you use `jq` to parse out the preferred version from `engines`, run it through the above check to get the best matching version, and then install that. It looks like you're already restoring a cached version of node and then [using `nvm` to install the desired version](https://github.com/netlify/build-image/blob/focal/run-build-functions.sh#L250).

If I'm reading this script correctly (totally possible I'm misunderstanding the use of the cache) then it almost seems like you're most of the way to having this particular, if less than ideal, solution working already without adding any extra lift for stuff like version range parsing. Maybe I'm wrong!

**Comment by noisy** [comment]
And looks like, that without this implemented, **I cannot deploy a new project after an update** to Netlify. This should be addressed ASAP. Everyday people have more and more reason to upgrade to newer version of node.

**Comment by nolessafool** [comment]
that's not true, @noisy .  You can use any node version you want; you need to set a node version in our way, which is use `$NODE_VERSION` and put whatever version you want in it, in the way we have documented here: https://docs.netlify.com/configure-builds/environment-variables/#netlify-configuration-variables

**Comment by MikeBeas** [comment]
@fool I mean, as demonstrated by that comment from @noisy, the fact that folks know about `engines` but not the Netlify-specific way to do this, and the fact that people even believe the only way to do this is `engines` should lend some credence to the idea that `engines` is the right way to do it and should be supported. 🤷🏻‍♂️ Just my take.

**Comment by adamreisnz** [comment]
I too, would appreciate it if support for `engines` is added, as it's extra work and easy to forget to update the Node version in Netlify config.

**Comment by chris-hailstorm** [comment]
Netlify people. Can you please explain why you aren’t fulfilling this
request? Using the engines key is the one, correct, completely obvious way
to implement this. The fact that this request has been open for years and
your customers are still asking you to do it should be concerning to you.
You are asking us to use a dumb, Netlify-specific approach.

Why?

**Comment by erezrokah** [comment]
Hi everyone and sorry for the slow responses on the issue. We are discussing internally the best way to approach it.

As mentioned the current workaround is to specify a custom Node.js version in order to opt out of the default one. We understand this is not ideal, and would like to suggest another possible solution.
Our default build image for new sites `focal`  includes Node.js version 16 which should support many use cases. If you're not using `focal` I recommend upgrading. See how in [our docs](https://docs.netlify.com/configure-builds/overview/#build-image-selection).

**Important:** After updating to `focal` an additional step is required to reset [pinned dependencies](https://docs.netlify.com/configure-builds/overview/#build-image-defaults-and-project-dependencies), which is re-linking your repo. There are instructions on doing do [here](https://docs.netlify.com/configure-builds/repo-permissions-linking/#unlink-a-git-repository) and [here](https://docs.netlify.com/configure-builds/repo-permissions-linking/#link-a-git-repository). We are also planning to remove the need for this additional step.

As for the issue at hand (detecting Node.js version from `package.json`), `package.json` engines field defines _**a range**_ and not a specific version. The purpose of the field is to describe the Node.js compatibility for consumers, and `npm/yarn` have specific behavior related to that (e.g. `yarn` will fail to install by default on non matching version, `npm` will only print a warning, unless [engine-strict](https://docs.npmjs.com/cli/v8/using-npm/config#engine-strict) is used). To summarize - it doesn't define which version should be used, it defines which versions are supported by the library.
Adding logic to infer a version based on it might create some confusion on which version the build will use. In order to provide the best experience it would be great to collect more information to make sure the behavior makes sense. The information we're looking for is:
1. The build version you're currently using (`xenial` or `focal`). You can get it from `Site settings > Build & deploy > Continuous Deployment > Build image selection`. It should also appear at the beginning of the deploy log.
2. The value of the engines field in your `package.json`
3. The Node.js version you expect us to use

Appreciate the feedback and looking forward to learning more about this.

>Another tip about `NODE_VERSION` is that it accepts any value supported by `nvm`, e.g. setting it to `node` will install the latest current version of Node.js which is `v17.8.0`

**Update**: added clarification on the purpose of the engines field.

**Comment by MikeBeas** [comment]
@erezrokah Howdy, thanks for the info. Glad to hear this is being discussed! Here's my input on those questions:

1. I'm on `focal`
2. My engines field is included below, used on two different Netlify sites
3. For this I would generally prefer the latest version of v17 available, but I would anticipate the possibility that it could be something as low as 17.0. If I wanted 17.8 or higher specifically I would set that version explicitly. (Same logic for npm.)

```json
{
  "engines": {
    "node": "17.x",
    "npm": "8.x"
  },
}
```

**Comment by erezrokah** [comment]
Thanks for the info @MikeBeas, quick follow up - do you mind sharing the reason Node.js 17 is required? What's missing from Node.js 16 that you need? I'm trying to figure out if 17 is a better default than 16

**Comment by MikeBeas** [comment]
@erezrokah I actually would recommend 16 as the default since that's the current LTS release. I only use 17 because I keep my personal installation up to date. I don't think I'm using anything specific from 17, it's just what I have installed so I use it on deployments as well to be safe.

**Comment by adamreisnz** [comment]
@fool @erezrokah here is another use case in favour of respecting the Node version in package.json rather than having to configure this through the dashboard.

We have a separate branch in our repositories called `maintenance`. We put this up when, as the name implies, we need to put our app in maintenance mode.

This branch however is severely outdated compared to the rest of the app. It's package.json contents refers to Node 12, whereas the main app is using Node 16.

This page and branch are very simple and only used several times a year at most. So there is no real benefit to us to update this branch to run and build on Node 16.

However, each time we deploy this branch, we now have to configure the Netlify app to use Node 12 instead of Node 16, or it won't build the maintenance branch. This is as you can imagine quite cumbersome.

It would be much preferred if Netlify could just read the `package.json` version for each branch, then use the appropriate Node version automatically, depending on which branch is being built.

**Comment by erezrokah** [comment]
Thanks for the additional use case @adamreisnz 👍 

So updating to a newer Node.js runtime version (from 12 to 16) causes the build to fail? Can you share a bit more information on the failure, for example the error you're getting when the build fail? The reason I'm asking is that we encourage people to upgrade to `focal` and if that can cause failures, it would be great to provide a migration guide.

Also, for your case using the following `netlify.toml` should make this a one time configuration:
```toml
[context.maintenance]
  environment = { NODE_VERSION = "12" }
```

More about this [here](https://docs.netlify.com/configure-builds/file-based-configuration/#sample-netlify-toml-file).

I agree this is not as automatic as it could be. Can you share the value of the engines field in your default branch and `maintenance` branch? Is it `"node": "16.x"` and `"node": "12.x"` accordingly?

**Comment by adamreisnz** [comment]
Yes, it does, it was a `gyp` related build failure. I don't have the exact logs, but can dig them up for you.
Fact of the matter is, as web developers can attest to, this happens more often than we'd like.

One node version up is usually not a problem, but as things progress, and you go up 2 or 3 node versions, and don't update packages or maintain a specific branch, chances are getting quite high that something will fail.

Normally, for your app, you will keep things up to date of course. But in our use case, for a maintenance branch which is rarely used, it was just not worth the hassle to do it.

In the end, we updated the branch to run on Node 16, but this was wasted time imo that could have been better spent elsewhere if Netlify could provide us with Node 12 for that specific branch build.

Glad that there is the `netlify.toml` configuration option, I will consider that for the next time.

But again I can't help but feel that this is a redundant duplication with the potential to cause errors, since the Node version is already listed in `package.json`

Our package.json versions are more specific, as we like to keep it up to date as newer versions become available.

Current app version is for example `"node": "^16.13.1"`, which means we want at least 16.13.1, but we would be happy with anything above that in the major version of Node 16.

The maintenance branch had something similar, but with Node 12. 

Hope this helps.

**Comment by coldino** [comment]
I don't want to be rude, but shouldn't this discussion stay on topic?

Talking about particular fail cases at one particular point in time may help one or a few people right now, but it does not help move towards a future-proof solution which matches developer expectations - something that supporting the `engine` field correctly would do.

I'm a big fan of the principle of least surprise. Having to hunt in the Netlify docs to set the NODE_VERSION variable is not hard by any means but it is surprising to need to do it at all when a well accepted standard solution already exists. So much of the Netlify experience is just smooth sailing that this just doesn't fit in.

**Comment by MentalGear** [comment]
I'm also requiring node 18 for the fetch API. (Ubuntu)
It would be great if we could use the standard-compliant "engines" property to set the node version to avoid having to setup yet another 3rd party file config.
